pkg/auth/auth_test.go

```diff
@@ -0,0 +1,390 @@
+	}
+}
+
+// TestAuthTokenFlag validates the use of --token will result in no
+// authentication flow being initialised, as well as the token not being
+// persisted to disk.
+func TestAuthTokenFlag(t *testing.T) {
+	wd, root := createTestEnvironment(t)
+	defer os.RemoveAll(root)
+	defer os.Chdir(wd)
+
+	var stdout bytes.Buffer
+	args := testutil.Args("pops --token 123")
+	opts := testutil.NewRunOpts(args, &stdout)
+	opts.ConfigPath = filepath.Join(root, manifest.Filename)
+	opts.ClientFactory = mock.ClientFactory(mockAPISuccess)
+
+	endpoint := make(chan string)
+	generatedToken := "123"
+	auth.Browser = mockBrowser(endpoint, generatedToken)
+	opts.AuthService = <-endpoint
+
+	err := app.Run(opts)
+	if err != nil {
+		t.Log(stdout.String())
+		t.Fatal(err)
+	}
+
+	wantToIgnore := "We are about to initialise a new authentication flow"
+	testutil.AssertStringDoesntContain(t, stdout.String(), wantToIgnore)
+
+	var f config.File
+	err = f.Read(opts.ConfigPath, opts.Stdin, opts.Stdout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if f.User.Token != "" {
+		t.Errorf("want no token, have '%s'", f.User.Token)
+	}
+	if f.User.Email != "" {
+		t.Errorf("want no email, have '%s'", f.User.Email)
+	}
+}
+
+// TestAuthMigration validates we're able to initialise the authentication flow
+// and to get a new access token returned and persisted to disk, which replaces
+// a pre-existing long-lived token (from before the CLI supported OAuth).
+func TestAuthMigration(t *testing.T) {
+	wd, root := createTestEnvironment(t)
+	defer os.RemoveAll(root)
+	defer os.Chdir(wd)
+
+	var stdout bytes.Buffer
+	args := testutil.Args("pops")
+	opts := testutil.NewRunOpts(args, &stdout)
+	opts.ConfigFile = config.File{
+		User: config.User{
+			Token: "123",
+		},
+	}
+	opts.ConfigPath = filepath.Join(root, manifest.Filename)
+	opts.ClientFactory = mock.ClientFactory(mockAPIMigration)
+	opts.Stdin = strings.NewReader("y") // Authorise opening of web browser.
+
+	endpoint := make(chan string)
+	generatedToken := "456"
+	auth.Browser = mockBrowser(endpoint, generatedToken)
+	opts.AuthService = <-endpoint
+
+	err := app.Run(opts)
+	if err != nil {
+		t.Log(stdout.String())
+		t.Fatal(err)
+	}
+
+	wantOutput := "Your current access token has no expiration and will be replaced with a short-lived token"
+	testutil.AssertStringContains(t, stdout.String(), wantOutput)
+
+	wantOutput = "We are about to initialise a new authentication flow"
+	testutil.AssertStringContains(t, stdout.String(), wantOutput)
+
+	var f config.File
+	err = f.Read(opts.ConfigPath, opts.Stdin, opts.Stdout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wantToken := "456" // token in config was 123 and should be replaced
+	if f.User.Token != wantToken {
+		t.Errorf("want token '%s', have '%s'", wantToken, f.User.Token)
+	}
+	wantEmail := "test@example.com"
+	if f.User.Email != wantEmail {
+		t.Errorf("want email '%s', have '%s'", wantEmail, f.User.Email)
+	}
+}
```

Comment:
Here is where we validate a successful authentication flow for a user who had a pre-existing token that was long-lived (i.e. not generated via the authentication service).